### PR TITLE
mgr/mds: add mds_partitioner

### PR DESCRIPTION
**Pull request by yongseokoh**
This PR provides mds_partitioner as mgr module for optimizing mds performance. Unlike the existing MDS Balancer, it employes ceph.dir.pin and ceph.dir.bal.mask to minimize metadata migration overhead. In addition, it considers the workload characteristics of each subtree and attempts to optimize performance by distributing it across all mds. It also provides two modes: manual mode, which is directly controlled by the administrator, and scheduler mode, which runs automatically and periodically.

Related PR: https://github.com/ceph/ceph/pull/52373

CDS Squid Discussion: https://pad.ceph.com/p/cds-squid-mds-partitioner-2023

tracker: https://tracker.ceph.com/issues/61778

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by yongseokoh** [comment]
jenkins retest this please

**Comment by yongseokoh** [comment]
jenkins test make check

**Comment by yongseokoh** [comment]
jenkins test api

**Comment by yongseokoh** [comment]
jenkins test dashboard cephadm

**Comment by yongseokoh** [comment]
jenkins test dashboard

**Comment by yongseokoh** [comment]
jenkins test api

**Comment by yongseokoh** [comment]
@vshankar Could you please review this PR?

**Comment by vshankar** [comment]
> @vshankar Could you please review this PR?

Absolutely. This change took a back seat for various reasons. I'll get it prioritized. Sorry for the delay!

**Comment by vshankar** [comment]
FYI - This is in my review queue for tomorrow.

**Comment by vshankar** [comment]
The nature of change is huge and will take some more time to finish completion.

**Review by vshankar — COMMENTED** [review]
I'll be done reviewing this by tomorrow. Nice work @yongseokoh 

This will be extremely beneficial once we have the changes baked and ready for use.

**Comment by vshankar** [comment]
I played with this today - interesting stuff. Nice work @yongseokoh 

Reviews coming up...

**Comment by yongseokoh** [comment]
@vshankar It could be very helpful to improve this PR.

**Comment by vshankar** [comment]
> I'll be done reviewing this by tomorrow. Nice work @yongseokoh

Almost done reviewing. Will publish my comments by EOD today.

File: src/pybind/mgr/mds_partitioner/fs/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    WAIT                 = 4
+    STOP                 = 5
+
+class Scheduler:
+    def __init__(self, analyzer, mover, policy_mgr, period=3600):
+        self.analyzer = analyzer
+        self.mover = mover
+        self.policy_mgr = policy_mgr
+        self.scheduler_period = period
+        self.analyzer_period = 30
```
Comment (vshankar) [inline src/pybind/mgr/mds_partitioner/fs/scheduler.py:27]: So, this is pretty much hard-coded and I should ask if 30s is enough for the analyser for the analyzer to decide on a partition?

**Review by vshankar — COMMENTED** [review]
Code organization is great. I've even tried playing with this along with the vxattr changes. Nice work @yongseokoh

File: src/pybind/mgr/mds_partitioner/fs/scheduler.py (hunk 1 above)
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/scheduler.py:27]: I think so. The time is fixed too short. Those values can be adjusted by parameter passing or ceph config set.

**Comment by yongseokoh** [comment]
> Code organization is great. I've even tried playing with this along with the vxattr changes. Nice work @yongseokoh

 I'm on vacation and have other things to do. I'll fix it in a few weeks. Thank you for the valuable feedback @vshankar.

File: src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+    def update_subtree_state(self, policy_name, subtree_path, state):
+        with self.lock:
+            if policy_name in self.workload_policies:
+                policy = self.workload_policies[policy_name]
+                policy.workload_table.update_subtree_state(subtree_path, state)
+                self._save(self.fs_name, policy_name, policy)
+
+    def activate(self, policy_name):
+        with self.lock:
+            if policy_name in self.workload_policies:
```
[the remark this answers is not in the document]
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py]: fixed

File: src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+                policy = self.workload_policies[policy_name]
+                policy.activate = True
+                self._save(self.fs_name, policy_name, policy)
+                return True, f"policy {policy_name} has been activated"
+
+            return False, f"{policy_name} is not available" 
+
+    def deactivate(self, policy_name):
+        with self.lock:
+            if policy_name in self.workload_policies:
```
[the remark this answers is not in the document]
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py]: fixed

File: src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+
+    def _remove(self, policy_name):
+        if policy_name not in self.workload_policies:
+            msg = f"policy_name {policy_name} doesn't exist."
+            log.debug(msg)
+            return False, msg
+
+        if self.workload_policies[policy_name].activate:
+            msg = f"{policy_name} cannot be removed because of activated."
+            log.debug(msg)
```
[the remark this answers is not in the document]
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/workload_policy_mgr.py]: removed

File: src/pybind/mgr/mds_partitioner/fs/perf_metric.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        fs_metric = self._get_stats()
+        if len(fs_metric['global_metrics'].get(fs_name, {})) == 0:
+            return perfs
+        client_to_dirpath = self._make_client_to_dirpath_map(fs_name, fs_metric, dirpath_list)
+        for metric_type in fs_metric_types:
+            perfs[metric_type.name] = self._collect_client_metric(fs_name, fs_metric, dirpath_list, client_to_dirpath, metric_type)
+        return perfs
+
+# to ms
+def calc_lat(c):
```
[the remark this answers is not in the document]
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/perf_metric.py:112]: Some comments have been added.

File: doc/cephfs/mds-partitioner.rst [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+To disable the mds_partitioner, run the following command::
+
+  ceph mds_partitioner disable $fs_name
+
+Command Line Interface
+======================
+
+mds_partitioner can be configured separately for each file system. 
+
+Enalbe/Disable mds_partitioner
```
[the remark this answers is not in the document]
Comment (yongseokoh) [reply doc/cephfs/mds-partitioner.rst]: Thank you

File: src/pybind/mgr/mds_partitioner/fs/scheduler.py (hunk 1 above)
Comment (yongseokoh) [reply src/pybind/mgr/mds_partitioner/fs/scheduler.py]: Parameters can be passed.
https://github.com/ceph/ceph/pull/54263/files#diff-5e25a13357d13efe4d4032b130722aaad8f750b216f6689dae5107497e74cd12R198

**Comment by yongseokoh** [comment]
@vshankar Modifications have been made based on your feedbacks.

**Pull request closed without merging**